### PR TITLE
Add `BloomForSequenceClassification` and `BloomForTokenClassification` classes

### PR DESCRIPTION
**Pull request by haileyschoelkopf**
# What does this PR do?

This PR adds 2 new classes for the BLOOM model with sequence classification and token classification heads. Mentioned briefly by me in PR #17474 .

We are planning to use the smaller BLOOM models for these tasks downstream in the Bigscience Multilingual Modeling WG, so we need these classes implemented to do so.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case). -> NO
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests? -> TODO

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 _Will tag patrickvonplaten and sgugger when ready--still need to write tests first!_

Let me know if anything is wrong with this PR!

### CODE REVIEW

**Comment by haileyschoelkopf** [comment]
@patrickvonplaten @sgugger

This PR should be ready for review. 

There are tests failing, but I am not sure why--I think I've looked at all the failing ones, and their failure seems to be unrelated to anything I changed (ie. image segmentation tests and tokenization tests fail, and text generation pipeline `test_small_model_pt` seems to fail because "sshleifer/tiny-ctrl" cannot be downloaded, none of which relate to files I touched in this PR afaik.)

File: src/transformers/models/bloom/modeling_bloom.py [hunk 1]
```diff
@@ -42,7 +47,7 @@
     "bigscience/bloom-1b3",
     "bigscience/bloom-2b5",
     "bigscience/bloom-6b3",
-    "bigscience/bloom-176b",
+    "bigscience/bloom",
```
Comment (sgugger) [inline src/transformers/models/bloom/modeling_bloom.py:50]: I'm not sure which one will be the final name, cc @younesbelkada

**Review by sgugger — APPROVED** [review]
LGTM once all tests pass!
We had an outage on the Hub last Friday and this weekend, relaunched all tests which should remove all flaky failures :-)

File: src/transformers/models/bloom/modeling_bloom.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_bloom.py:50]: Indeed the final model is `bloom`and not `bloom-176b` ! Thanks for spotting the nit :)

**Comment by younesbelkada** [comment]
Hi @haileyschoelkopf , I have managed to fix some tests that fails on this PR, can I directly push to this PR in case the tests still fails on your side?

**Comment by younesbelkada** [comment]
Perfect thanks! Just pushed my commit

**Comment by haileyschoelkopf** [comment]
After a minor change to return type of `BloomForTokenClassification` , all tests pass! 

This should be ready to merge now :)

File: src/transformers/models/bloom/modeling_bloom.py [hunk 2]
```diff
@@ -959,3 +964,223 @@ def _reorder_cache(past: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor) ->
...
+    Named-Entity-Recognition (NER) tasks.
+    """,
+    BLOOM_START_DOCSTRING,
+)
+class BloomForTokenClassification(BloomPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [r"h.*.self_attention.scale_mask_softmax.causal_mask", r"lm_head.weight"]
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
```
Comment (younesbelkada) [inline src/transformers/models/bloom/modeling_bloom.py:1110]: Should we add a default value for `num_labels` in the config?

Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_bloom.py:1110]: My bad, by default it is set to `2` so it should be fine!    https://github.com/haileyschoelkopf/transformers/blob/1d5914033d482c4ecc810213f0bc68e9d10b3c86/src/transformers/configuration_utils.py#L316

**Review by younesbelkada — APPROVED** [review]
Left a very small comment! Otherwise looks very good to me! Thank you very much for helping us implementing these classes

**Comment by haileyschoelkopf** [comment]
Comments should be resolved now!

**Comment by stefan-it** [comment]
Hi @haileyschoelkopf ,

thanks for adding this! Do you have any recommendations for a good hyper-param configuration (batch size, learning rate) when doing NER? I tried CoNLL-2003 with the `bigscience/bloom-350m` checkpoint, but results are around 67% on test set, which is very bad.

**Comment by haileyschoelkopf** [comment]
Hi @stefan-it, I agree with what @younesbelkada said, bloom-1b3 is worth trying since it's finished pretraining! We haven't actually gotten to NER experiments but I'll let you know if we do end up finding good hyperparams.
